### PR TITLE
SONAR-9616 Support notifications on branches

### PR DESCRIPTION
**Pull request by janos-ss**

### CODE REVIEW

**Comment by janos-ss** [comment]
I will add some ITs tomorrow.

File: server/sonar-server/src/main/java/org/sonar/server/projectbranch/ws/BranchDtoToWsBranch.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+import static org.sonar.db.component.BranchType.SHORT;
+import static org.sonar.server.projectbranch.ws.BranchesWs.DEFAULT_MAIN_BRANCH_NAME;
+
+public class BranchDtoToWsBranch {
+
+  private BranchDtoToWsBranch() {
+    // static methods only
+  }
+
+  static WsBranches.Branch.Builder toBranchBuilder(BranchDto branch, Optional<BranchDto> mergeBranch, Map<String, MeasureDto> measuresByMetricKey) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/projectbranch/ws/BranchDtoToWsBranch.java]: MINOR Specify a "BranchDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

File: server/sonar-db-dao/src/main/java/org/sonar/db/component/ComponentDto.java [hunk 2]
```diff
@@ -441,4 +497,39 @@ public String toString() {
...
+    copy.language = language;
+    copy.description = description;
+    copy.tags = tags;
+    copy.enabled = enabled;
+    copy.isPrivate = isPrivate;
+    copy.createdAt = createdAt;
+    return copy;
+  }
+
+  // TODO Use it in branch plugin
```
Comment (SonarTech) [inline server/sonar-db-dao/src/main/java/org/sonar/db/component/ComponentDto.java]: INFO Complete the task associated to this TODO comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)

File: server/sonar-db-dao/src/main/java/org/sonar/db/component/ComponentDto.java [hunk 3]
```diff
@@ -317,6 +358,10 @@ public ComponentDto setDescription(@Nullable String description) {
     return this;
   }
 
+  /**
+   * Use {@link #projectUuid()}, {@link #moduleUuid()} or {@link #moduleUuidPath()}
+   */
+  @Deprecated
   public String getRootUuid() {
```
Comment (SonarTech) [inline server/sonar-db-dao/src/main/java/org/sonar/db/component/ComponentDto.java]: MAJOR Add the missing @deprecated Javadoc tag. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)
INFO Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

File: server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentAction.java [hunk 4]
```diff
@@ -137,6 +153,16 @@ private ComponentWsResponse doHandle(ComponentWsRequest request) {
     }
   }
 
+  private ComponentDto loadComponent(DbSession dbSession, ComponentWsRequest request) {
+    String componentKey = request.getComponent();
+    String componentId = request.getComponentId();
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentAction.java]: MINOR Remove this use of "getComponentId"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentAction.java [hunk 5]
```diff
@@ -242,8 +268,9 @@ private static void addBestValuesToMeasures(List<MeasureDto> measures, Component
 
   private static ComponentWsRequest toComponentWsRequest(Request request) {
     ComponentWsRequest componentWsRequest = new ComponentWsRequest()
-      .setComponentId(request.param(PARAM_COMPONENT_ID))
-      .setComponentKey(request.param(PARAM_COMPONENT_KEY))
+      .setComponentId(request.param(DEPRECATED_PARAM_COMPONENT_ID))
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentAction.java]: MINOR Remove this use of "setComponentId"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: server/sonar-server/src/main/java/org/sonar/server/component/ComponentCleanerService.java [hunk 6]
```diff
@@ -51,11 +52,15 @@ public void delete(DbSession dbSession, List<ComponentDto> projects) {
     }
   }
 
+  public void deleteBranch(DbSession dbSession, ComponentDto branch) {
+    // TODO: detect if other branches depend on it?
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/component/ComponentCleanerService.java]: INFO Complete the task associated to this TODO comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/QualityGateEventsStep.java [hunk 7]
```diff
@@ -25,6 +25,8 @@
 import org.sonar.api.notifications.Notification;
 import org.sonar.api.utils.log.Logger;
 import org.sonar.api.utils.log.Loggers;
+import org.sonar.server.computation.task.projectanalysis.analysis.AnalysisMetadataHolder;
+import org.sonar.server.computation.task.projectanalysis.analysis.Branch;
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/QualityGateEventsStep.java]: MINOR Remove this unused import 'org.sonar.server.computation.task.projectanalysis.analysis.Branch'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/issue/ComponentIssuesLoader.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    this.ruleRepository = ruleRepository;
+  }
+
+  public List<DefaultIssue> loadForComponentUuid(String componentUuid) {
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      List<DefaultIssue> result = new ArrayList<>();
+      dbSession.getMapper(IssueMapper.class).scrollNonClosedByComponentUuid(componentUuid, resultContext -> {
+        DefaultIssue issue = (resultContext.getResultObject()).toDefaultIssue();
+
+        // TODO this field should be set outside this class
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/issue/ComponentIssuesLoader.java]: INFO Complete the task associated to this TODO comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/issue/ComponentIssuesLoader.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        if (!isActive(issue.ruleKey()) || ruleRepository.getByKey(issue.ruleKey()).getStatus() == RuleStatus.REMOVED) {
+          issue.setOnDisabledRule(true);
+          // TODO to be improved, why setOnDisabledRule(true) is not enough ?
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/issue/ComponentIssuesLoader.java]: INFO Complete the task associated to this TODO comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/issue/ComponentIssuesLoader.java [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+          issue.setBeingClosed(true);
+        }
+        // FIXME
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/issue/ComponentIssuesLoader.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1134)

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SendIssueNotificationsStep.java [hunk 11]
```diff
@@ -22,10 +22,12 @@
 import com.google.common.collect.ImmutableSet;
 import java.util.Date;
 import java.util.Map;
+import java.util.Optional;
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SendIssueNotificationsStep.java]: MINOR Remove this unused import 'java.util.Optional'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: server/sonar-db-dao/src/main/java/org/sonar/db/purge/PurgeConfiguration.java [hunk 12]
```diff
@@ -34,24 +35,27 @@
   private final IdUuidPair rootProjectIdUuid;
   private final String[] scopesWithoutHistoricalData;
   private final int maxAgeInDaysOfClosedIssues;
+  private final Optional<Integer> maxAgeInDaysOfInactiveShortLivingBranches;
   private final System2 system2;
   private final Collection<String> disabledComponentUuids;
 
   public PurgeConfiguration(IdUuidPair rootProjectId, String[] scopesWithoutHistoricalData, int maxAgeInDaysOfClosedIssues,
-    System2 system2, Collection<String> disabledComponentUuids) {
+    Optional<Integer> maxAgeInDaysOfInactiveShortLivingBranches, System2 system2, Collection<String> disabledComponentUuids) {
```
Comment (SonarTech) [inline server/sonar-db-dao/src/main/java/org/sonar/db/purge/PurgeConfiguration.java]: MINOR Specify a "Integer" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

File: server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentTreeAction.java [hunk 13]
```diff
@@ -275,8 +289,9 @@ private static ComponentTreeWsRequest toComponentTreeWsRequest(Request request)
     List<String> metricKeys = request.mandatoryParamAsStrings(PARAM_METRIC_KEYS);
     checkArgument(metricKeys.size() <= MAX_METRIC_KEYS, "Number of metrics keys is limited to %s, got %s", MAX_METRIC_KEYS, metricKeys.size());
     ComponentTreeWsRequest componentTreeWsRequest = new ComponentTreeWsRequest()
-      .setBaseComponentId(request.param(PARAM_BASE_COMPONENT_ID))
-      .setBaseComponentKey(request.param(PARAM_BASE_COMPONENT_KEY))
+      .setBaseComponentId(request.param(DEPRECATED_PARAM_BASE_COMPONENT_ID))
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentTreeAction.java]: MINOR Remove this use of "setBaseComponentId"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/TreeAction.java [hunk 14]
```diff
@@ -270,6 +288,7 @@ private static TreeWsRequest toTreeWsRequest(Request request) {
     TreeWsRequest treeWsRequest = new TreeWsRequest()
       .setBaseComponentId(request.param(PARAM_COMPONENT_ID))
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/TreeAction.java]: MINOR Remove this use of "setBaseComponentId"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/TreeAction.java [hunk 15]
```diff
@@ -182,6 +190,16 @@ private TreeWsResponse doHandle(TreeWsRequest treeWsRequest) {
     }
   }
 
+  private ComponentDto loadComponent(DbSession dbSession, TreeWsRequest request) {
+    String componentId = request.getBaseComponentId();
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/TreeAction.java]: MINOR Remove this use of "getBaseComponentId"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/TreeAction.java [hunk 16]
```diff
@@ -270,6 +288,7 @@ private static TreeWsRequest toTreeWsRequest(Request request) {
     TreeWsRequest treeWsRequest = new TreeWsRequest()
       .setBaseComponentId(request.param(PARAM_COMPONENT_ID))
       .setBaseComponentKey(request.param(PARAM_COMPONENT))
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/TreeAction.java]: MINOR Remove this use of "setBaseComponentKey"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/TreeAction.java [hunk 17]
```diff
@@ -182,6 +190,16 @@ private TreeWsResponse doHandle(TreeWsRequest treeWsRequest) {
     }
   }
 
+  private ComponentDto loadComponent(DbSession dbSession, TreeWsRequest request) {
+    String componentId = request.getBaseComponentId();
+    String componentKey = request.getBaseComponentKey();
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/TreeAction.java]: MINOR Remove this use of "getBaseComponentKey"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/component/ComponentKeyGenerator.java [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.computation.task.projectanalysis.component;
+
+import javax.annotation.Nullable;
+import org.sonar.scanner.protocol.output.ScannerReport;
+
+@FunctionalInterface
+public interface ComponentKeyGenerator {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/component/ComponentKeyGenerator.java]: MAJOR Drop this interface in favor of "java.util.function.BiFunction<ScannerReport$Component,ScannerReport$Component,String>". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1711)

File: server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentTreeDataLoader.java [hunk 19]
```diff
@@ -128,6 +130,16 @@ ComponentTreeData load(ComponentTreeWsRequest wsRequest) {
     }
   }
 
+  private ComponentDto loadComponent(DbSession dbSession, ComponentTreeWsRequest request) {
+    String componentKey = request.getComponent();
+    String componentId = request.getBaseComponentId();
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentTreeDataLoader.java]: MINOR Remove this use of "getBaseComponentId"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/issue/TrackerMergeBranchInputFactory.java [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+  private final ComponentIssuesLoader mergeIssuesLoader;
+  private final DbClient dbClient;
+  private final AnalysisMetadataHolder analysisMetadataHolder;
+  private Map<String, String> uuidsByKey;
+
+  public TrackerMergeBranchInputFactory(ComponentIssuesLoader mergeIssuesLoader, AnalysisMetadataHolder analysisMetadataHolder, DbClient dbClient) {
+    this.mergeIssuesLoader = mergeIssuesLoader;
+    this.analysisMetadataHolder = analysisMetadataHolder;
+    this.dbClient = dbClient;
+    // TODO detect file moves?
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/issue/TrackerMergeBranchInputFactory.java]: INFO Complete the task associated to this TODO comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/issue/TrackerMergeBranchInputFactory.java [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+  }
+
+  private void loadMergeBranchComponents() {
+    String mergeBranchUuid = analysisMetadataHolder.getBranch().get().getMergeBranchUuid().get();
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/issue/TrackerMergeBranchInputFactory.java]: MAJOR Call "Optional#isPresent()" before accessing the value. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3655)

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/BuildComponentTreeStep.java [hunk 22]
```diff
@@ -64,64 +67,52 @@ public String getDescription() {
...
-    private final DbSession dbSession;
-    private String projectUuid = null;
-    private Optional<SnapshotDto> cache = null;
-
-    private BaseAnalysisSupplier(DbClient dbClient, DbSession dbSession) {
-      this.dbClient = dbClient;
-      this.dbSession = dbSession;
-    }
+  private ComponentKeyGenerator loadKeyGenerator() {
+    return Stream.of(analysisMetadataHolder.getBranch(), Optional.of(new DefaultKeyGenerator()))
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/BuildComponentTreeStep.java]: MAJOR Call "Optional#isPresent()" before accessing the value. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3655)

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/BuildComponentTreeStep.java [hunk 23]
```diff
@@ -64,64 +67,52 @@ public String getDescription() {
...
+      // TODO pull request generator will be added here
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/BuildComponentTreeStep.java]: INFO Complete the task associated to this TODO comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 27 issues
* ![MAJOR][MAJOR] 8 major
* ![MINOR][MINOR] 12 minor
* ![INFO][INFO] 7 info

Watch the comments in this conversation to review them.

#### 3 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [ComponentTreeAction.java#L315](https://github.com/SonarSource/sonarqube/blob/da772d3f4d5cb235f00e9a3d553ad883c5db3ce7/server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentTreeAction.java#L315): A "NullPointerException" could be thrown; "getSort" is nullable here. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259)
1. ![MAJOR][MAJOR] [ComponentTreeAction.java#L316](https://github.com/SonarSource/sonarqube/blob/da772d3f4d5cb235f00e9a3d553ad883c5db3ce7/server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentTreeAction.java#L316): A "NullPointerException" could be thrown; "getSort" is nullable here. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259)
1. ![MAJOR][MAJOR] [ComponentTreeAction.java#L321](https://github.com/SonarSource/sonarqube/blob/da772d3f4d5cb235f00e9a3d553ad883c5db3ce7/server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentTreeAction.java#L321): A "NullPointerException" could be thrown; "getSort" is nullable here. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/QualityGateEventsStep.java [hunk 24]
```diff
@@ -25,6 +25,7 @@
 import org.sonar.api.notifications.Notification;
 import org.sonar.api.utils.log.Logger;
 import org.sonar.api.utils.log.Loggers;
+import org.sonar.server.computation.task.projectanalysis.analysis.AnalysisMetadataHolder;
 import org.sonar.server.computation.task.projectanalysis.component.Component;
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/QualityGateEventsStep.java]: MINOR Remove this unused import 'org.sonar.server.computation.task.projectanalysis.analysis.Branch'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [SendIssueNotificationsStep.java#L25](https://github.com/SonarSource/sonarqube/blob/93849ac690d182d6c8f76998653a25fcd2dba356/server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/SendIssueNotificationsStep.java#L25): Remove this unused import 'java.util.Optional'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: server/sonar-server/src/test/java/org/sonar/server/computation/task/projectanalysis/step/QualityGateEventsStepTest.java [hunk 25]
```diff
@@ -71,11 +73,13 @@
   private MeasureRepository measureRepository = mock(MeasureRepository.class);
   private EventRepository eventRepository = mock(EventRepository.class);
   private NotificationService notificationService = mock(NotificationService.class);
-  private QualityGateEventsStep underTest = new QualityGateEventsStep(treeRootHolder, metricRepository, measureRepository, eventRepository, notificationService);
+  private AnalysisMetadataHolder analysisMetadataHolder = mock(AnalysisMetadataHolder.class);
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/computation/task/projectanalysis/step/QualityGateEventsStepTest.java]: You could have used AnalysisMetadataHolderRule, it allows to not use mock

**Pull request merged**